### PR TITLE
CMake 3.18+ Device Link Xcompiler option fix for Warnings as Errors

### PR DESCRIPTION
**Pull request by ptheywood**
Fix cmake related issue for passing multiple options to xcompiler at device linking.

In the process of opening a cmake bug about the unexpected behaviour.

### CODE REVIEW

**Comment by Robadob** [comment]
CMake 3.16.1, VS 2015 is giving me this on configure (multiple times)

```
CMake Warning (dev) at cmake/common.cmake:287:
  Syntax Warning in cmake code at column 93

  Argument not separated from preceding token by whitespace.
Call Stack (most recent call first):
  src/CMakeLists.txt:26 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.
```

File: cmake/common.cmake [hunk 1]
```diff
@@ -277,14 +277,14 @@ function(add_flamegpu_executable NAME SRC FLAMEGPU_ROOT PROJECT_ROOT IS_EXAMPLE)
...
+        if(WARNINGS_AS_ERRORS)               
+            if(CMAKE_CXX_COMPILER_ID STREQUAL "MSVC")
+                target_link_options(${NAME} PRIVATE "$<DEVICE_LINK:SHELL:-Xcompiler /WX>")
             else()
-                target_link_options(${NAME} PRIVATE $<DEVICE_LINK:-Xcompiler /wd4100>)
-            endif()
-        else()
-            if(WARNINGS_AS_ERRORS)
-                target_link_options(${NAME} PRIVATE $<DEVICE_LINK:-Xcompiler -Werror>)
+                target_link_options(${NAME} PRIVATE "$<DEVICE_LINK:SHELL:-Xcompiler -Werror">)
```
Comment (Robadob) [inline cmake/common.cmake:287]: Final quote need to be moved right.

**Comment by ptheywood** [comment]
> Does not build under visual studio 2015 with Werror on.
> 
> e.g. main lib files `LayerDescription.cpp`,`ModelData.cpp`, `AgentData.cpp`, `AgentFunctionData.cpp`, `AgentDescription.cpp`, `AgentFunctionDescription.cpp`, `curve_rtc.cpp` have these errors (all the same error, the include to blame isn't immediately obvious):
> 
> ```
> 1>C:\Program Files (x86)\Windows Kits\8.1\Include\um\DbgHelp.h(1544): error C2220: warning treated as error - no 'object' file generated (compiling source file C:\Users\Robadob\Desktop\fgpu2\src\flamegpu\runtime\cuRVE\curve_rtc.cpp)
> 1>C:\Program Files (x86)\Windows Kits\8.1\Include\um\DbgHelp.h(1544): warning C4091: 'typedef ': ignored on left of '' when no variable is declared (compiling source file C:\Users\Robadob\Desktop\fgpu2\src\flamegpu\runtime\cuRVE\curve_rtc.cpp)
> 1>C:\Program Files (x86)\Windows Kits\8.1\Include\um\DbgHelp.h(3190): warning C4091: 'typedef ': ignored on left of '' when no variable is declared (compiling source file C:\Users\Robadob\Desktop\fgpu2\src\flamegpu\runtime\cuRVE\curve_rtc.cpp)
> 1
> ```
> 
> e.g. gtest has these reorder errors
> 
> ```
> 4>C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\xlocbuf(324): error : the initialization of member "std::wstring_convert<_Codecvt, _Elem, _Walloc, _Balloc>::_Has_state [with _Codecvt=wchar_codecvt, _Elem=__wchar_t, _Walloc=std::allocator<__wchar_t>, _Balloc=std::allocator<char>]" will be done before that of member "std::wstring_convert<_Codecvt, _Elem, _Walloc, _Balloc>::_Has_werr [with _Codecvt=wchar_codecvt, _Elem=__wchar_t, _Walloc=std::allocator<__wchar_t>, _Balloc=std::allocator<char>]"
> 4>            detected during instantiation of "std::wstring_convert<_Codecvt, _Elem, _Walloc, _Balloc>::wstring_convert() [with _Codecvt=wchar_codecvt, _Elem=__wchar_t, _Walloc=std::allocator<__wchar_t>, _Balloc=std::allocator<char>]"
> 4>  C:\Users\Robadob\Desktop\fgpu2\build\tests\googletest-src\googletest\include\gtest/internal/gtest-port.h(2060): here
> 4>
> ```

I have neither Windows kit 8.1, nor VS2015 installed so can't reproduce these (would rather not have to wait several hours to install both and reinstall CUDA etc.)

Given the `C4091` is from an include named `DbgHelp` I assume this is debug mode only, which I probably didn't test, really not clear what header / change must be causing that. 
I assume this can't be specific to this branch either - as those are certainly not device link related. 

The `wreorder` issue should dissapear thanks to one of the other PRs, which removes  Wreorder from MSVC.

**Comment by Robadob** [comment]
Yeah it's not the end of the world, being vs2015 (it does work in less strict mode). I'd like to see how it performs after the other branches are merged though. So i'll wait on this one.

**Comment by ptheywood** [comment]
@Robadob to test this locally and if happy it can be merged.

**Pull request merged**